invenio_app_ils/anonymization.py

```diff
@@ -0,0 +1,177 @@
+from invenio_circulation.api import Loan
+from invenio_circulation.proxies import current_circulation
+from invenio_db import db
+from invenio_oauthclient.models import RemoteAccount, UserIdentity
+from invenio_userprofiles.models import UserProfile
+
+from invenio_app_ils.acquisition.api import Order
+from invenio_app_ils.acquisition.proxies import current_ils_acq
+from invenio_app_ils.document_requests.api import DocumentRequest
+from invenio_app_ils.errors import PatronNotFoundError
+from invenio_app_ils.ill.api import BorrowingRequest
+from invenio_app_ils.ill.proxies import current_ils_ill
+from invenio_app_ils.patrons.api import (Patron, get_anonymous_patron_dict,
+                                         get_patron_or_unknown)
+
+from .acquisition.search import OrderSearch
+from .circulation.search import get_loans_by_patron_pid
+from .document_requests.search import DocumentRequestSearch
+from .ill.search import BorrowingRequestsSearch
+from .patrons.indexer import PatronIndexer
+from .proxies import current_app_ils
+
+
+def get_patron_activity(patron_pid):
+    """Get activity related to the given patron pid."""
+    if patron_pid is None:
+        raise ValueError("No patron pid was provided.")
+
+    patron = get_patron_or_unknown(patron_pid)
+    if not patron:
+        return None
+
+    def dump(search):
+        return [hit.to_dict() for hit in search.scan()]
+
+    patron_document_requests = dump(
+        DocumentRequestSearch().search_by_patron_pid(patron_pid)
+    )
+
+    patron_borrowing_requests = dump(
+        BorrowingRequestsSearch().search_by_patron_pid(patron_pid)
+    )
+
+    patron_acquisitions = dump(OrderSearch().search_by_patron_pid(patron_pid))
+
+    patron_loans = dump(get_loans_by_patron_pid(patron_pid))
+
+    patron_profile = UserProfile.get_by_userid(patron_pid).__dict__
+
+    patron_data = {
+        "patron": patron,
+        "profile": patron_profile,
+        "document_requests": patron_document_requests,
+        "borrowing_requests": patron_borrowing_requests,
+        "acquisitions": patron_acquisitions,
+        "loans": patron_loans,
+    }
+
+    return patron_data
+
+
+def anonymize_patron_data(patron_pid, force=False):
+    """Anonymize all the data/activity related to the given patron pid."""
+    if patron_pid is None:
+        raise ValueError("No patron pid was provided.")
+
+    patron = get_patron_or_unknown(patron_pid)
+    if not patron and not force:
+        return None
+
+    patron_object = None
+    try:
+        patron_object = current_app_ils.patron_cls.get_patron(patron_pid)
+    except PatronNotFoundError:
+        if not force:
+            raise PatronNotFoundError(patron_pid)
+
+    # Delete rows from db
+    dropped = 0
+
+    with db.session.begin_nested():
+        ongoing_loans = 0
+        for loan in (
+            get_loans_by_patron_pid(patron_pid)
+            .filter("term", state="ITEM_ON_LOAN")
+            .scan()
+        ):
+            ongoing_loans += 1
```

Comment:
Just out of curiosity, there wasn't any other way to retrieve the total amount of results?

Reply:
Yes! 
It can be retrieved using ES aggregations and it is faster.
It would look something like this:
```
loan_states = (
            current_app.config["CIRCULATION_STATES_LOAN_ACTIVE"]
        )
loans_result = get_loans_by_patron_pid(patron_pid, loan_states)[:0].execute()
        for bucket in loans_result.aggregations.states.buckets:
            if bucket["key"] in current_app.config["CIRCULATION_STATES_LOAN_ACTIVE"]:
                ongoing_loans += bucket["doc_count"]
```
For the states it is better to retrieve it from the config.
And would also need to change the `get_loans_by_patron_pid` to something like this:
```
def get_loans_by_patron_pid(patron_pid, states):
    """Returns all the loans (past and current) for a given patron."""
    search = current_circulation.loan_search_cls()\
        .filter("terms", state=states)\
        .filter("term", patron_pid=patron_pid)
    # Aggregation
    aggs = A("terms", field="state")
    search.aggs.bucket("states", aggs)
    return search
```

Reply:
@jrcastro2 nice! Just a last question, are we sure with the aggregations all the results are being returned to that count? I know that scan is used when you want to retrieve at once more than 10000 results (I think that is the maximum result window). Also, another solution I was thinking is: Normal search and then just retrieve the total count, not sure if that has limitations in the number of results though...

Reply:
Thanks to you!
Regarding your first comment: we assumed that any given user could only have at most a few active loans at any given time, therefore our decision to scan the results instead of aggregating them.
We will still go with Javier's good suggestion. It does not matter much if the number of retrieved results is limited because it's probably not going to happen in practice, and even if it did the function would not break (we are only interested in knowing if the user has at least one active loan, the exact count is not really important).